src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlBufferReader.cs

```diff
@@ -436,10 +423,21 @@ public unsafe decimal ReadDecimal()
             int offset;
             byte[] buffer = GetBuffer(ValueHandleLength.Decimal, out offset);
             decimal value;
-            byte* pb = (byte*)&value;
-            for (int i = 0; i < sizeof(decimal); i++)
-                pb[i] = buffer[offset + i];
-            Advance(ValueHandleLength.Decimal);
+            if (BitConverter.IsLittleEndian)
+            {
+                byte* pb = (byte*)&value;
+                for (int i = 0; i < sizeof(decimal); i++)
+                    pb[i] = buffer[offset + i];
+                Advance(ValueHandleLength.Decimal);
+            }
+            else
+            {
+                int* pi = (int*)&value;
+                pi[0] = GetInt32(offset + 12); // flags
+                pi[1] = GetInt32(offset + 8); // hi32
+                pi[2] = GetInt32(offset); // bottom-of-low64;
+                pi[3] = GetInt32(offset + 4); // top-of-low64;
```

Comment:
This looks wrong, which probably explains the failures in the decimal tests on s390x.

Upstream code uses the decimal-from-int-array constructor, which uses a different ordering compared to the in-memory layout.  Is there any reason for not just backporting this from upstream as-is?

If you do want to use the "int *" unsafe method, the destination needs to match the definition of the decimal type:
```
        private readonly int _flags;
        private readonly uint _hi32;
        private readonly ulong _lo64;
```
so we have `flags`, `hi32`, `top-of-low64`, `bottom-of-low64` in that order on a BE host system.

On the other hand, the *source* also seems to be wrong: this has to match the LE in-memory layout of the decimal type, i.e. `flags`, `hi32`, `bottom-of-low64`, `top-of-low64`.   Overall, it looks like this should be:
```
                pi[0] = GetInt32(offset); // flags
                pi[1] = GetInt32(offset + 4); // hi32
                pi[2] = GetInt32(offset + 12); // top-of-low64;
                pi[3] = GetInt32(offset + 8); // bottom-of-low64;
```
unless I'm missing something ...